File: Baekjoon/yerim/카드 구매하기.cpp
// [11052] 카드 구매하기
#include <iostream>
using namespace std;

int n, p[1001], dp[1001];

void input()
{
    cin >> n;
    for(int i = 1; i <= n; i++) {
        cin >> p[i];
    }
}

void getMaxCost()
{
    dp[0] = 0; dp[1] = p[1];
    for(int i = 2; i <= n; i++) {
        for(int j = i; j >= 1; j--) {
            dp[i] = max(dp[i], p[j] + dp[i - j]);
        }
    }
}

int main(void)
{
    ios_base::sync_with_stdio(false);
    cin.tie(NULL);

    input();
    getMaxCost();
    cout << dp[n] << '\n';

    return 0;
}

File: Baekjoon/subin/카드 구매하기.py
n = int(input())
cards = [0] + list(map(int, input().split()))

d = [0] * (n + 1)

for i in range(1, n + 1):
    for j in range(1, i + 1):
        d[i] = max(d[i], d[i - j] + cards[j])

print(d[-1])